### ISSUE TITLE
Why does  LLaVA-v1.5-7B XTuner get higher accuracy than the official

### ISSUE THREAD

**Issue by mileistone** [InternLM/xtuner#352]
![image](https://github.com/InternLM/xtuner/assets/10510712/27fb0422-d667-42cc-8816-515e1a381276)

I find that LLaVA-v1.5-7B XTuner gets higher accuracy than the official, may I ask what is the difference between XTuner and official?

**Comment by LZHgrla** [comment]
@mileistone  Hi!

We found that the official implementation of `sampler` did not synchronize seeds between multiple ranks. 
In this situation, some data samples were actually trained multiple times, while others were trained 0 time.

Other parts are completely consistent with the official implementation.

**Comment by mileistone** [comment]
@LZHgrla Thank you！Your work is great. LLaVA-InternLM2-7B using around 558k pairs to pre-train achieves comparable performance with InternLMX-Compser which is pre-trained on 1.4B image-caption pairs.

**Comment by LZHgrla** [comment]
Oh, thanks very much!

There is another modification. We utilize the `prompt_template` during the pre-train, while official implementation doesn't.

https://github.com/InternLM/xtuner/blob/f225761eeaa9e3c63704db04a0252e67792c8eb0/xtuner/configs/llava/vicuna_7b_v15_clip_vit_large_p14_336/pretrain/llava_vicuna_7b_v15_clip_vit_large_p14_336_e1_gpu8_pretrain.py#L31

This modification will not bring significant accuracy improvement, but in our opinion, adding the `prompt_template` for chat model is "politically correct".

**Comment by mileistone** [comment]
@LZHgrla LLaVA official seems also utilize prompt_template https://github.com/haotian-liu/LLaVA/blob/main/llava/conversation.py#L58.
![image](https://github.com/InternLM/xtuner/assets/10510712/94a88926-16fe-48c4-bd81-224944871aa4) 
Do I misunderstand something?

**Comment by mileistone** [comment]
@LZHgrla I agree with you. In my experiment, whether we have prompt_template in pre-train seems has no effect on MMBench.

**Comment by shipengai** [comment]
> but in our opinion

@LZHgrla 哪些代码是解决这个问题的？

**Comment by shipengai** [comment]
> @mileistone Hi!
> 
> We found that the official implementation of `sampler` did not synchronize seeds between multiple ranks. In this situation, some data samples were actually trained multiple times, while others were trained 0 time.
> 
> XTuner fixes this bug.
> 
> Other parts are completely consistent with the official implementation.

@LZHgrla  哪些代码是解决这个问题的？

**Comment by shipengai** [comment]
> > > @mileistone Hi!
> > 
> > 
> > > 
> > 
> > 
> > > We found that the official implementation of `sampler` did not synchronize seeds between multiple ranks. In this situation, some data samples were actually trained multiple times, while others were trained 0 time.
> > 
> > 
> > > 
> > 
> > 
> > > XTuner fixes this bug.
> > 
> > 
> > > 
> > 
> > 
> > > Other parts are completely consistent with the official implementation.
> > 
> > 
> > @LZHgrla  哪些代码是解决这个问题的？
> 
> https://github.com/InternLM/xtuner/blob/f63859b3d0cb39cbac709e3850f3fe01de1023aa/xtuner/dataset/samplers/length_grouped.py#L125
> 
> Use a consistent random generator for different ranks

@LZHgrla 关于seed这段代码，我们在同一台机器4卡测试发现，并没有影响。  你们改的这段代码只有在多机多卡情况下才有影响吧。

**Comment by LZHgrla** [comment]
@shipengai 
感谢您的测试，先前我对LLaVA的实现理解有误，您是正确的。

在最近的code review中，我们发现了又一处实现的不同，即每个 group 的数据样本大小。
LLaVA 使用了 `world_size * batch_size`

https://github.com/haotian-liu/LLaVA/blob/5d8f1760c08b7dfba3ae97b71cbd4c6f17d12dbd/llava/train/llava_trainer.py#L91

https://github.com/haotian-liu/LLaVA/blob/5d8f1760c08b7dfba3ae97b71cbd4c6f17d12dbd/llava/train/llava_trainer.py#L71


而 xtuner 的实现参考了transformers的group_sampler，在此基础上额外增加了一个系数 `mega_batch_mult`，以确保同一iter尽可能同样的长度。

https://github.com/InternLM/xtuner/blob/c7c97b9a64bb1915e4e212f46f2f74f934d9564b/xtuner/dataset/samplers/length_grouped.py#L109

https://github.com/huggingface/transformers/blob/5c341d4555ba3e4b656053317e372ebed0c5af37/src/transformers/trainer_pt_utils.py#L545

**Comment by shipengai** [comment]
@LZHgrla 感谢你的回复。
看了[https://github.com/huggingface/transformers/blob/5c341d4555ba3e4b656053317e372ebed0c5af37/src/transformers/trainer_pt_utils.py#L533C5-L534C82](url)   的介绍，相当于首先对训练数据根据token的长度，进行了排序，第一iter是由最长的训练数据按照最开始的batch_size来构成。因此，在训练时候，在第一个iter和最后一个iter，用的batch_size是不一样的，这样理解对吗？

**Comment by LZHgrla** [comment]
@shipengai 
在我看来，此处的意思是将最长的数据放到第一个 batch 中，这样可以保证 OOM 出现在训练开始阶段，而非训练中间阶段。所以这个地方，应该是与 batchsize 无关

**Comment by shipengai** [comment]
@LZHgrla 所以xtuner相较于官方代码实现的性能提升只来源于训练数据过程中的采样不同？

**Comment by LZHgrla** [comment]
@shipengai 
从目前已掌握的情况来看，是这样子的。
